### ISSUE TITLE
Add 0/1 swizzle syntax

### ISSUE THREAD

**Issue by dneto0** [gpuweb/gpuweb#732]
From https://twitter.com/FreyaHolmer/status/1255107846129299456

```
okay I swear to god, the amount of time I spend doing this:

float3( v.xy, 0 )
float3( v.x, 0, v.y )
float4( v, 0 )
float4( v, 1 )

like, if only we had swizzle operators for 0 and 1, imagine if we could just

v.xy0
v.x0y
v.xyz0
v.xyz1
```

So, the idea is to permit `0` and `1` as elements of a swizzle syntax, where they indicate not positions in the source vector, but rather constants 0 and 1 (of the right scalar type).

At LLVM or SPIR-V level, there's no disadvantage, as a general swizzle to construct a vector use the "shuffle" operation, and in this case the second operand to the shuffle would be a 0-1 vector rather than a dummy (unused) vector.

**Comment by jdashg** [comment]
Oh my goodness yes. I have always been frustrated that GLSL didn't have this, and it seems so easy to do.

I suppose the ambiguity to resolve is whether (or how) we support `v.0yzw`, since this would require something different from `expr DOT ident` in the grammar.

**Comment by magcius** [comment]
As much as I like this feature, I think it would make sense to defer until the HLSL team has evaluated such a feature ( filed as issue https://github.com/microsoft/DirectXShaderCompiler/issues/2699 ). If HLSL team uses a different form of the same syntax (e.g. choosing `_01xy` to more closely align with matrices, which can also start their swizzle tokens with digits, albeit for indices in that case), that could hurt the readability of WGSL as developers switch between the two languages.

**Comment by litherum** [comment]
This seems like a good idea. The ratio of author benefit to implementation cost seems quite high.

If we do https://github.com/gpuweb/gpuweb/issues/737, then these `v.x0y` should be considered as rvalues.

**Comment by johnkslang** [comment]
I just want to add a few notes of how this might by trickier than it first appears, though that also depends on how other parts of WGSL are defined, so I'll try to treat those as dualisms:

**1. Constant folding and propagation.**

One area of portability concern is having front ends not do this exactly the same way.  With 100s of operators/built-ins, there is room for error in which subset must be folded and must not be folded (you don't want grey area in between).

So, consider these (pseudo code):

```
vec4Var._1;  // does this become a compile time 1.0f, or remain an expression?
// if a compile-time 1, then
<expression>._1  // this is also a compile-time 1.0f, but is expression evaluated?
// if expression has a side effect, it should be evaluated
<undefined-behavior-expression>._1 // can a smart front end see that toss the expression?
```

Have such questions been answered for integer multiply by 0?

HLLs have portability issues directly related to such expressions and to which subset of expressions must fold and propagate constants.

And, in the floating-point space, questions about precision differences between the host and the target, such that two compiles running on the same device can get different answers.

**2. Implementation in an HLL via a constructor, to avoid widening the output interface**

```
vec4Var._1yx;  // to implement as a constructor, you want...
vec3(1., vec4Var.y, vec4Var.x);  // that's okay, but note the replication of vec4Var
<expression>._1yx;  // will you replicate expression?
vec3(1., <expression>.y, <expression>.x); // okay if expression is small and has no side effect
```

Obviously, you want to add a copy, though it's interesting to note that GLSL through glslang never has to do this for any other situation (it does for HLSL).  Basically, it requires an algorithm like:

1. Is it a symbol? if yes, reuse it.
1. Does it have a side effect?  If yes, execute it, then replicate the result.
2. Is it a very simple expression?  If yes, replicate it.
3. Otherwise, execute it, then replicate the result.

While almost obviously simple (but not quite, someone will make a mistake somewhere), the point is, every other operation is representable in glslang's AST without such an algorithm.  And, it's a bit lossy, unfortunately, and would be nicer to capture this was the request.

Conclusion:  Front-ends are better off with native AST or output IR support of this operation.

That is, while it appears to be syntactic sugar, it may be best to handle as widening the pipe between all layers of the stack.

**3. Swizzling scalars**

Will you support this?

```
float x;
x._1;  // a scalar 1.f (side subject: there's a required type conversion to whatever the vector type is)
```

It comes up when mechanically (or even mind-numbingly) making a function/operator to handle "all four widths" of type, in the absence of true generics.  E.g., I want:

```
vec4Var._xxx1;
vec3Var._xx1;
vec2Var._x1;
scalarVar._1
```

Without the scalar case, everyone has to special case the final line.  Hence GLSL added the convenience of supporting it.  If you want the first 3 lines but not the 4th line, what's the argument? The 4th line is just for convenience, so we don't need it, and the first 3 lines are so convenient we need them?

With the scalar case, you get to pin down types/expression/constants/etc. as per the rest of the discussion.

**Summary**

If WGSL has no constant propagation (good for portability), it will not be a real language for actually creating complex shaders, but a good intermediate language, and you'd use an existing HLL to target WGSL, and we don't even need to talk about such convenience features, they're way down on the list below having simple constant folding of constant expressions.

If WGSL does have good constant folding, A) you have to pin down the issues above, and B) it's again playing in the HLL space of difficult portability, with a less familiar syntax, begging the question of why not start with an existing HLL, reduce it a bit, add what's need, and fix it where we have already discovered what the portability issues are, instead of making a whole new set of unknowns.  (I'm currently in the camp of making a low-level highly portable WGSL to be used like an IR.)

**Comment by litherum** [comment]
Can we just forbid those cases?

> `vec4Var._1`
> `scalarVar._1`

**Comment by johnkslang** [comment]
Yes, you could.  Forbidding scalar.x, scalar._1, etc. is consistent with an HLL that has many features with many limitations and needing a complex negative testing system to make sure no one accidentally implements the obvious generalities and orthogonalities, thereby creating portability problems.

A lot of GLSL portability issues have a root cause in a vendor implementing better generality/orthogonality than the specification allowed for

Have a small number of features that are required in their full generality and orthogonality is consistent with a lower-level language that is easier to port and test.  You can spot test the positive combinatorics to make sure it is all there, as opposed to exhaustively testing the negative combinatorics of the edge cases to make sure none got implemented.

I believe we are after the small/simple/general/orthogonal/portable/testable path, rather than the  lots of partial features path that make it easier to write code but then contradicts all the simple/general/orthogonal/portable/testable goals.

**Comment by litherum** [comment]
I'm so confused. You just brought up a bunch of problems with doing things like `vec4Var._1`. It seems natural we should forbid things which have lots of problems.

We can cover these cases in tests, requiring them to be forbidden.

**Comment by johnkslang** [comment]
My point was to be cautious adding this feature, because it seems you don't want to add it all the way, which is a root cause of portability issues with HLLs.  The grey areas outside the favorite idiomatic parts to support get interpreted slightly different ways by different compilers.

**Comment by grorg** [comment]
Very briefly discussed at the [2020-06-09 meeting](https://docs.google.com/document/d/1ap37K-4SaybRS9Qdt7fj8SNkwOnoH3GyVBPKLYDMm8s/).

**Comment by skal65535** [comment]
any hope for this one to move forward? What was the "good point it’s deceptively complex" by John?

I'd love to see this one happen, happens sooo often to me.
